### PR TITLE
feat: FFT v1

### PR DESCRIPTION
**Pull request by ShubyM**
## Summary

This PR adds a first working full fine-tuning (FFT) path for OpenRL where two SFT jobs can share one local GPU through explicit scheduler handoff.

The main shape is:

- full-FT runs get their own trainer worker process
- each worker drains only its own run queue
- workers acquire/release CUDA access through a node-local trainer scheduler
- the scheduler uses `cuda-checkpoint` to checkpoint a worker on release and restore it before the next acquire
- the GSM8K SFT recipe demonstrates two concurrent full-FT jobs against the same gateway

For the detailed design notes, see [`AGENTS.md`](./AGENTS.md).

This prototype is mainly controlled by two server-side knobs:

- `OPEN_RL_TRAINING_MODE=full`
  - Forces training-client creation through the full fine-tuning path on the server.
  - This lets the stock Tinker client/cookbook loop keep calling the existing LoRA-shaped API while the server creates a full-FT worker.

- `OPEN_RL_SCHEDULER_SOCKET=/tmp/open-rl/trainer-scheduler.sock`
  - Enables scheduled full-FT workers.
  - The gateway launches a per-run worker, and workers coordinate CUDA access through this Unix socket.

`examples/sft/gsm8k` contains the demo path:

- start the scheduler
- start the gateway with full-FT mode and the scheduler socket
- launch two GSM8K SFT jobs with different `log_path`s
- use the printed `eval_model_path=...` for post-training vLLM eval

Sampling during training is still out of scope for this PR; eval is a separate post-training step.

Image show the scheduler checkpointing between two different jobs (note the different PID's)
image

### CODE REVIEW

File: examples/sft/gsm8k/README.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# GSM8K full fine-tuning (via the tinker server)
+
+Full-parameter SFT of a small model on [GSM8K](https://github.com/openai/grade-school-math)
+(7,473 grade-school math problems with chain-of-thought), driven through the open-rl gateway
+with the tinker SDK.
+
+## Why full-FT goes through an env var
+
+The tinker SDK's `create_lora_training_client()` is LoRA-only (it rejects `training_mode`). To do
```
Comment (droot) [inline examples/sft/gsm8k/README.md:9]: Note this PR: we should investigate if create_lora_training_client accept untyped dict that we leverage to provide an extra arg ...

File: examples/sft/gsm8k/README.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+## Run (4-Pane Setup)
+
+This setup demonstrates multi-tenant full fine-tuning (FFT) SFT jobs sharing a single GPU through explicit scheduler handoff. 
+
+Assumes you start at the project root in all terminal panes:
+
+### Pane 1: Scheduler
+```bash
+uv --project src/server run python src/server/trainer_scheduler.py
```
Comment (droot) [inline examples/sft/gsm8k/README.md:22]: Scheduler is an overloaded term and let's avoid it. Especially in context of OpenRL, scheduler is expected to be a global component responsible for scheduling the training/sampling work.

In this case, `trainer_scheduler.py` is suppose to be running on each node irrespective of trainer or sampler. May be we should simply call this: snapshot_agent and it will align well with the time-slicing workstream to avoid any confusion.

File: examples/sft/gsm8k/README.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+## Eval
+
+Eval is decoupled (point a server at the saved dir). With vLLM (run from the project root):
+
+```bash
+python examples/sft/gsm8k/vllm_eval.py \
+  --path <saved_dir> --data gsm8k_test.json
+```
+
+## Results (Qwen2.5-0.5B, 1 epoch, lr 2e-5, 0-shot exact-match on 250 test problems)
```
Comment (droot) [inline examples/sft/gsm8k/README.md:65]: so interleaving produces the same result as without interleaving ?

File: src/server/checkpoint.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import os
```
Comment (droot) [inline src/server/checkpoint.py:1]: lets create a separate component called `snapshot-agent` and move the functionality of checkpoint/restore to that directory.

The OpenRL component, the trainer worker, will use the snapshot-agent's client to communicate with snapshot agent.

File: src/server/gateway.py [hunk 5]
```diff
@@ -51,7 +53,8 @@ def filter(self, record: logging.LogRecord) -> bool:
 
 
 def is_single_process_mode() -> bool:
-  return bool(os.getenv("BASE_MODEL")) and not bool(os.getenv("REDIS_URL"))
+  # A shared store (Redis or a file-backed dir) means workers run as separate processes.
+  return bool(os.getenv("BASE_MODEL")) and not (os.getenv("REDIS_URL") or os.getenv("OPEN_RL_STORE_DIR"))
```
Comment (droot) [inline src/server/gateway.py:57]: didn't understand this part. what does `OPEN_RL_STORE_DIR` means here ? it is responsible for storing what ?

File: src/server/gateway.py [hunk 6]
```diff
@@ -77,6 +80,43 @@ def sampler_weights_path(model_id: str, name: str) -> str:
...
+def local_weights_path(model_id: str, name: str) -> str:
+  return os.path.join(TMP_DIR, "checkpoints", model_id, "weights", name)
+
+
+def resolve_weights_path(path: str, model_id: str | None = None) -> str:
+  if os.path.isabs(path):
+    return path
+
+  parsed = parse_tinker_path(path)
+  if parsed is not None and parsed[1] == "weights":
```
Comment (droot) [inline src/server/gateway.py:105]: We can always assume tinker style path even for local weights, right ?

File: src/server/gateway.py [hunk 7]
```diff
@@ -188,7 +228,7 @@ async def client_config(_: dict):
 
 @app.post("/api/v1/create_session")
 async def create_session(_: dict):
-  return {"session_id": "sess-real-123", "type": "create_session"}
+  return {"session_id": str(uuid.uuid4()), "type": "create_session"}
```
Comment (droot) [inline src/server/gateway.py:231]: Should we persist it ?

May be we should add a TODO for session persistence.

File: src/server/gateway.py [hunk 8]
```diff
@@ -202,14 +242,32 @@ async def create_model(req: dict):
   base_model = req.get("base_model")
   if not base_model:
     return JSONResponse(status_code=400, content={"error": "base_model is required"})
-  model_id = str(uuid.uuid4())
+  # OPEN_RL_TRAINING_MODE=full forces every job to full fine-tuning, so the stock (LoRA-only)
+  # tinker SDK create_lora_training_client() yields a full-FT model with no client changes.
+  force_full = os.getenv("OPEN_RL_TRAINING_MODE", "").lower() == "full"
+  training_mode = "full" if force_full or req.get("training_mode") == "full" or req.get("full_finetune") or req.get("is_lora") is False else "lora"
```
Comment (droot) [inline src/server/gateway.py:248]: req.get("training_mode") is there way to do that today ? if yes, then we should make that completely client driven...right ?

Also req.get("full_finetune") ... why so many ways to do the same thing ?

File: AGENTS.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
+# Full Fine-Tuning Scheduler Diff
```
Comment (droot) [inline AGENTS.md:1]: AGENTS.md file is for managing the dev on the repo itself but it looks like we have updated it to focus just on full finetuning stuff.

should this be designs/fft.md ?

File: AGENTS.md [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+- Only one worker can hold the scheduler's active CUDA slot at a time.
+- A worker must call `ACQUIRE` before CUDA work and `RELEASE` after its drain window.
+- `RELEASE` checkpoints the active worker with `cuda-checkpoint`.
+- A later `ACQUIRE` restores that same worker if it was checkpointed.
+
+The intended first milestone is narrow: two full fine-tuning SFT jobs, same base model, one OpenRL server, one GPU residency slot. This PR does not try to solve sampling, vLLM handoff, multi-base scheduling, placement, multi-GPU topology, or crash recovery.
+
+## Design Shape
+
+The scheduler uses the same basic shape as `accel-time-slicer`:
```
Comment (droot) [inline AGENTS.md:28]: looks like this is referencing the accel-time-slicer project that is internal.

File: src/server/trainer_scheduler.py [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+from typing import Any
+
+from checkpoint import CheckpointRestorer, CudaCheckpointRestorer
+from utils import timed
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class Worker:
```
Comment (droot) [inline src/server/trainer_scheduler.py:20]: s/Worker/Process ?

File: src/server/trainer_scheduler.py [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+  async def register(self, run_id: str, pid: int) -> dict[str, Any]:
+    return await self.request({"command": "REGISTER", "run_id": run_id, "pid": pid})
+
+  async def unregister(self, run_id: str) -> dict[str, Any]:
+    return await self.request({"command": "UNREGISTER", "run_id": run_id})
+
+  @asynccontextmanager
+  async def acquire(self, run_id: str) -> AsyncIterator[None]:
+    await self.request({"command": "ACQUIRE", "run_id": run_id})
```
Comment (droot) [inline src/server/trainer_scheduler.py:209]: When we test with bigger models, we will run in to timeouts on these connections. In my original implementation, I had to add retries and explicit ACK to make the logic more robust.

Basically, robust thing will to implement async API but I am fine not doing it right now because we will swap it out with snapshot-agent anyways.

But it will be good if we can try testing M1 on bigger model to see how it holds sooner rather than later.

File: src/server/store.py [hunk 13]
```diff
@@ -177,6 +218,90 @@ async def get_future(self, req_id: str, timeout: float) -> dict[str, Any] | None
     return {"type": "try_again", "request_id": req_id, "queue_state": "active"}
 
 
+class FileStore(RequestStore):
+  """Cross-process store backed by a shared directory — no external server (Redis) needed.
```
Comment (droot) [inline src/server/store.py:222]: I think it's totally fine to assume dependency on Redis running locally and not introduce too many variables in the picture. Testing different store implementations is pain in the long run.

**Review by droot — COMMENTED** [review]
Took a first pass, will take another look later.

File: src/server/gateway.py (hunk 5 above)
Comment (droot) [reply src/server/gateway.py:57]: like we discussed, lets use redis and call it a day.

File: src/server/gateway.py [hunk 14]
```diff
@@ -77,6 +80,43 @@ def sampler_weights_path(model_id: str, name: str) -> str:
   return f"tinker://{model_id}/sampler_weights/{name}"
 
 
+def weights_path(model_id: str, name: str) -> str:
```
Comment (droot) [inline src/server/gateway.py:83]: it will be good to add a comment explaining what does this path represent and how it is different from the `sampler_weights_path` above ? should this be named as `trainer_weights_path` ?

It will also be good to explain what does `sampler_session_id` means and what is the significance of appending seq_id ?

File: src/server/gateway.py [hunk 15]
```diff
@@ -77,6 +80,43 @@ def sampler_weights_path(model_id: str, name: str) -> str:
...
+  return f"tinker://{model_id}/weights/{name}"
+
+
+def parse_tinker_path(path: str) -> tuple[str, str, str] | None:
```
Comment (droot) [inline src/server/gateway.py:87]: not clear to me... what path is this expecting ? sampler ? session ? trainer ?

File: src/server/gateway.py [hunk 16]
```diff
@@ -77,6 +80,43 @@ def sampler_weights_path(model_id: str, name: str) -> str:
...
+    return path
+
+  parsed = parse_tinker_path(path)
+  if parsed is not None and parsed[1] == "weights":
+    return local_weights_path(parsed[0], parsed[2])
+
+  if model_id is not None:
+    return local_weights_path(model_id, path)
+
+  return os.path.join(TMP_DIR, "checkpoints", path)
```
Comment (droot) [inline src/server/gateway.py:111]: all these weights paths are getting out of hand ...

File: src/server/gateway.py [hunk 17]
```diff
@@ -202,14 +242,32 @@ async def create_model(req: dict):
   base_model = req.get("base_model")
   if not base_model:
     return JSONResponse(status_code=400, content={"error": "base_model is required"})
-  model_id = str(uuid.uuid4())
+  # OPEN_RL_TRAINING_MODE=full forces every job to full fine-tuning, so the stock (LoRA-only)
+  # tinker SDK create_lora_training_client() yields a full-FT model with no client changes.
+  force_full = os.getenv("OPEN_RL_TRAINING_MODE", "").lower() == "full"
+  training_mode = "full" if force_full or req.get("training_mode") == "full" or req.get("full_finetune") or req.get("is_lora") is False else "lora"
+  session_id = req.get("session_id")
+  model_seq_id = req.get("model_seq_id")
```
Comment (droot) [inline src/server/gateway.py:250]: it will be good to point out what does this represent.

File: src/server/gateway.py [hunk 18]
```diff
@@ -202,14 +242,32 @@ async def create_model(req: dict):
...
+  if not session_id:
+    return JSONResponse(status_code=400, content={"error": "session_id is required"})
+  if model_seq_id is None:
+    return JSONResponse(status_code=400, content={"error": "model_seq_id is required"})
+  model_id = f"{session_id}-{model_seq_id}"
+  if training_mode == "full" and (scheduler_socket := os.getenv("OPEN_RL_SCHEDULER_SOCKET")):
+    try:
+      process = launch_worker(model_id, scheduler_socket)
```
Comment (droot) [inline src/server/gateway.py:258]: Not clear to me: what is the dependency on scheduler_socket ? when launching training_worker...

We should think how it will work in distributed system ?

File: src/server/gateway.py [hunk 19]
```diff
@@ -202,14 +242,32 @@ async def create_model(req: dict):
...
+
   req_id = await _enqueue(
     {
       "req_id": model_id,
       "model_id": model_id,
       "type": "create_model",
       "base_model": base_model,
       "lora_config": req.get("lora_config") or {},
+      "training_mode": training_mode,
     }
```
Comment (droot) [inline src/server/gateway.py:271]: I am wondering if api server launching the trainer worker is the right design here... should this be done by the request processor part....

File: src/server/gateway.py [hunk 20]
```diff
@@ -299,12 +389,6 @@ async def optim_step(req: dict):
 
 @app.post("/api/v1/save_weights_for_sampler")
 async def save_weights_for_sampler(req: dict):
-  """TrainingClient.save_weights_for_sampler().
-
-  The SDK uses this for both named sampler checkpoints and ephemeral
-  save_weights_and_get_sampling_client() snapshots. Route it through the trainer
-  queue so the sampler always sees weights saved after prior training requests.
-  """
   model_id = req.get("model_id")
```
Comment (droot) [inline src/server/gateway.py:392]: that comment was useful pl. bring it back.

File: src/server/gateway.py [hunk 21]
```diff
@@ -327,20 +412,14 @@ async def save_weights_for_sampler(req: dict):
...
 @app.post("/api/v1/save_weights")
 async def save_weights(req: dict):
-  """TrainingClient.save_weights() / save_state() — persists adapter to TMP_DIR/checkpoints/<alias>.
-
-  This is the endpoint the tinker SDK hits for both save_weights() and save_state().
-  When `path` is provided we treat it as a named checkpoint alias and resolve to
-  TMP_DIR/checkpoints/<alias> (or leave absolute paths alone), so subsequent
-  `create_training_client_from_state(alias)` calls can find the adapter.
-  """
   model_id = req.get("model_id")
```
Comment (droot) [inline src/server/gateway.py:415]: I see that function comment was useful. 

AGENTS.md should probably have instructions saying... don't remove existing comments and also don't make unnecessary changes.

File: src/server/store.py (hunk 13 above)
Comment (droot) [reply src/server/store.py:222]: like we discussed, let's remove this pl.

File: src/server/store.py [hunk 22]
```diff
@@ -155,6 +176,26 @@ async def get_requests(self) -> list[dict[str, Any]]:
 
     return batch
 
+  async def get_requests_for_model(self, model_id: str) -> list[dict[str, Any]]:
```
Comment (droot) [inline src/server/store.py:179]: Not clear to me... why was this change introduced ?

File: src/server/trainer.py [hunk 23]
```diff
@@ -17,6 +17,10 @@
 ENABLE_GRADIENT_CHECKPOINTING = os.getenv("ENABLE_GRADIENT_CHECKPOINTING", "1") == "1"
 
 
+def defer_cuda_probe() -> bool:
```
Comment (droot) [inline src/server/trainer.py:20]: No idea... what this is and how is it used and why it has to be linked with OPEN_RL_SCHEDULER

File: src/server/trainer.py
Comment (droot) [inline src/server/trainer.py:51]: What is the relationship of this with trainer-worker ?

File: src/server/trainer.py [hunk 24]
```diff
@@ -79,11 +84,33 @@ def load_base_model(self, base_model_name: str) -> None:
...
     self.base_model_name = base_model_name
     self.tokenizer = AutoTokenizer.from_pretrained(base_model_name)
-    dtype = torch.bfloat16 if torch.cuda.is_bf16_supported() else torch.float32
+    dtype = torch.bfloat16 if self.device.type == "cuda" and torch.cuda.is_bf16_supported() else torch.float32
 
     self.base_model = AutoModelForCausalLM.from_pretrained(base_model_name, dtype=dtype, device_map=self.device)
     print("Successfully loaded.")
 
+  def create_full_model(self, model_id: str, base_model_name: str) -> None:
+    """Materialize one full-parameter fine-tuning model for this worker."""
```
Comment (droot) [inline src/server/trainer.py:93]: what does materialize means here ?..... put it on vram (if yes, would be good to state that this needs to be invoked with gpu acquired) ?

File: src/server/trainer.py [hunk 25]
```diff
@@ -564,12 +690,10 @@ def generate(
...
-    assert self.peft_model is not None, "Model must be loaded first."
-
+  def _prompt_logprobs(self, input_tensor: torch.Tensor, model: Any) -> list[float | None]:
     with torch.no_grad():
       attention_mask = torch.ones_like(input_tensor)
-      outputs = self.peft_model(input_tensor, attention_mask=attention_mask)
+      outputs = model(input_tensor, attention_mask=attention_mask)
       logprob_dist = torch.nn.functional.log_softmax(outputs.logits[0, :-1], dim=-1)
 
     prompt_tokens = input_tensor[0].tolist()
```
Comment (droot) [inline src/server/trainer.py:699]: I tried reviewing this file, but had very hard time understanding it honestly. LoRA and full feels like super mixed in this implementation.
